src/models/participant/participantCached.js

```diff
@@ -0,0 +1,141 @@
+const ParticipantModel = require('../../models/participant/participant')
+
+let cacheClient
+let participantsAllCacheKey
+
+/*
+  Private API
+*/
+
+const buildUnifiedParticipantsData = (allParticipants) => {
+  // build indexes (or indices?) - optimization for byId and byName access
+  const indexById = {}
+  const indexByName = {}
+
+  allParticipants.forEach((oneParticipant) => {
+    // Participant API returns Date type, but cache internals will serialize it to String
+    // by calling JSON.stringify(), which calls .toISOString() on a Date object.
+    // Let's ensure all places return same kind of String.
+    oneParticipant.createdDate = JSON.stringify(oneParticipant.createdDate)
+
+    // Add to indexes
+    indexById[oneParticipant.participantId] = oneParticipant
+    indexByName[oneParticipant.name] = oneParticipant
+  })
+
+  // build unified structure - indexes + data
+  const unifiedParticipants = {
+    indexById,
+    indexByName,
+    allParticipants
+  }
+
+  return unifiedParticipants
+}
+
+const getParticipantsCached = async () => {
+  // Do we have valid participants list in the cache ?
+  let cachedParticipants = cacheClient.get(participantsAllCacheKey)
+  if (!cachedParticipants) {
+    // No participants in the cache, so fetch from participan API
+    const allParticipants = await ParticipantModel.getAll()
+    cachedParticipants = buildUnifiedParticipantsData(allParticipants)
+
+    // store in cache
+    cacheClient.set(participantsAllCacheKey, cachedParticipants)
+  } else {
+    // unwrap participants list from catbox structure
+    cachedParticipants = cachedParticipants.item
+  }
+  return cachedParticipants
+}
+
+/*
+  Public API
+*/
+exports.initialize = async () => {
+  /* Register as cache client */
+  const participantCacheClientMeta = {
+    id: 'participants',
+    preloadCache: getParticipantsCached
+  }
+
+  cacheClient = Cache.registerCacheClient(participantCacheClientMeta)
+  participantsAllCacheKey = cacheClient.createKey('all')
+}
+
+exports.invalidateParticipantsCache = async () => {
+  cacheClient.drop(participantsAllCacheKey)
+}
+
+exports.getById = async (id) => {
+  try {
+    const cachedParticipants = await getParticipantsCached()
+    return cachedParticipants.indexById[id]
+  } catch (err) {
+    throw ErrorHandler.Factory.reformatFSPIOPError(err)
+  }
+}
+
+exports.getByName = async (name) => {
+  try {
+    const cachedParticipants = await getParticipantsCached()
+    return cachedParticipants.indexByName[name]
+  } catch (err) {
+    throw ErrorHandler.Factory.reformatFSPIOPError(err)
+  }
+}
+
+exports.getAll = async () => {
+  try {
+    const cachedParticipants = await getParticipantsCached()
+    return cachedParticipants.allParticipants
+  } catch (err) {
+    throw ErrorHandler.Factory.reformatFSPIOPError(err)
+  }
+}
+
+const withInvalidate = (theFunctionName) => {
+  return async (...args) => {
+    try {
+      const result = await ParticipantModel[theFunctionName](...args)
```

Comment:
Should you not use the cache mechanism to retrieve the data once the cache has been invalidated?